src/onemod/pipeline.py

```diff
@@ -175,6 +209,96 @@ def stage_from_json(
             stage_class = getattr(module, stage_type)
         return stage_class.from_json(filepath, name, from_pipeline)
 
+    def build_dag(self) -> dict[str, list[str]]:
+        """Build directed acyclic graph (DAG) from the stages and their dependencies."""
+        # TODO: Placeholder until DAG class is implemented, assuming we need one
+        return self._dependencies
+
+    def validate_dag(self):
+        """Validate that the DAG structure is correct."""
+        for stage, dependencies in self._dependencies.items():
+            # Check for undefined dependencies
+            for dep in dependencies:
+                if dep not in self._stages:
+                    raise ValueError(
+                        f"Stage '{dep}' is not defined, but '{stage}' depends on it."
+                    )
+
+            # Check for self-dependencies
+            if stage in dependencies:
+                raise ValueError(f"Stage '{stage}' cannot depend on itself.")
+
+            # Check for duplicate dependencies
+            if len(dependencies) != len(set(dependencies)):
+                raise ValueError(
+                    f"Duplicate dependencies found for stage '{stage}'."
+                )
+
+        # Check for isolated nodes
+        # TODO: are there cases where isolated nodes would be valid for OneMod?
+        all_stages = set(self._stages.keys())
+        dependent_stages = set(
+            stage for deps in self._dependencies.values() for stage in deps
+        )
+        isolated_stages = (
+            all_stages - dependent_stages - set(self._dependencies.keys())
+        )
+        if isolated_stages:
+            logger.warning(
+                f"The following stages are isolated and not part of the DAG: {isolated_stages}"
+            )
+
+    def get_execution_order(self) -> list[str]:
+        """
+        Return topologically sorted order of stages, ensuring no cycles.
+        Uses of Kahn's algorithm to find the topological order of the stages.
+        """
+        # Track in-degrees of all nodes
+        in_degree = {stage: 0 for stage in self._stages}
+
+        # Populate in-degree based on dependencies
+        for stage, dependencies in self._dependencies.items():
+            for dep in dependencies:
+                in_degree[stage] += 1
```

Comment:
Should this be `in_degree[dep]`?

Reply:
Oops, no, the in-degree is the number of incoming nodes, i.e., the number of dependencies a node has, right? Could we use `in_degree = {stage: len(dependencies) for stage, dependencies in self._dependencies.items()}` in line 257? I'm assuming all stages are in `self._dependencies`. If not, could we use `in_degree[stage] = len(dependencies)` in line 261?

Reply:
We can. The first suggestion would mean we are relying on `_dependencies` only, rather than initializing the set of stages based on `_stages`, which is probably fine, as that should probably be validated elsewhere (I actually think they may not need to be separate entities ultimately, but this is a larger design question that we are in the process of answering). Either way, will update so the method is less busy.